In [1]:
#import necessary libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
import statsmodels.api as sm

In [4]:
#read file
df = pd.read_csv("/content/1. Bankruptcy.csv")
df.head()

,FYEAR,Bankruptcy Status,CUSTOMERID,Ratio1,Ratio2,Ratio3,Ratio4,Ratio5,Ratio6,Ratio7,Ratio8,Ratio9,Ratio10
0,1999,0,36020,0.307139,0.887006,1.647681,-0.199158,1.092964,-0.313289,-0.196793,1.206763,0.282471,0.158896
1,1999,0,36110,0.760736,0.592493,0.453003,-0.369890,0.186154,0.039619,0.327497,0.428418,1.106965,0.793443
2,1999,0,37520,-0.513596,0.337615,0.299015,-0.029080,-0.432605,0.829993,-0.707786,0.476153,2.179175,2.484584
3,1994,1,78110,-0.466129,0.370747,0.496067,-0.373429,-0.267424,0.977799,-0.610975,0.456810,0.151951,0.047789
4,1999,0,00079X10,2.023422,0.214876,0.182595,6.695360,-1.148338,-1.505889,2.876477,0.287375,-0.986442,0.791077


In [5]:
df.info() # Display basic information about the DataFrame
df.isnull().sum() # Check for missing values in each column

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2675 entries, 0 to 2674
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   FYEAR              2675 non-null   int64  
 1   Bankruptcy Status  2675 non-null   int64  
 2   CUSTOMERID         2675 non-null   object 
 3   Ratio1             2675 non-null   float64
 4   Ratio2             2675 non-null   float64
 5   Ratio3             2675 non-null   float64
 6   Ratio4             2675 non-null   float64
 7   Ratio5             2675 non-null   float64
 8   Ratio6             2675 non-null   float64
 9   Ratio7             2675 non-null   float64
 10  Ratio8             2675 non-null   float64
 11  Ratio9             2675 non-null   float64
 12  Ratio10            2675 non-null   float64
dtypes: float64(10), int64(2), object(1)
memory usage: 271.8+ KB


,0
FYEAR,0
Bankruptcy Status,0
CUSTOMERID,0
Ratio1,0
Ratio2,0
Ratio3,0
Ratio4,0
Ratio5,0
Ratio6,0
Ratio7,0


In [6]:
df = df.drop('CUSTOMERID', axis=1) # Drop the 'CUSTOMERID' column as it is an identifier and not useful for modeling
display(df.head()) # Display the first 5 rows of the updated DataFrame to verify the column removal

,FYEAR,Bankruptcy Status,Ratio1,Ratio2,Ratio3,Ratio4,Ratio5,Ratio6,Ratio7,Ratio8,Ratio9,Ratio10
0,1999,0,0.307139,0.887006,1.647681,-0.199158,1.092964,-0.313289,-0.196793,1.206763,0.282471,0.158896
1,1999,0,0.760736,0.592493,0.453003,-0.369890,0.186154,0.039619,0.327497,0.428418,1.106965,0.793443
2,1999,0,-0.513596,0.337615,0.299015,-0.029080,-0.432605,0.829993,-0.707786,0.476153,2.179175,2.484584
3,1994,1,-0.466129,0.370747,0.496067,-0.373429,-0.267424,0.977799,-0.610975,0.456810,0.151951,0.047789
4,1999,0,2.023422,0.214876,0.182595,6.695360,-1.148338,-1.505889,2.876477,0.287375,-0.986442,0.791077


In [7]:
#Scaling Financial Ratios
from sklearn.preprocessing import StandardScaler

# Select the ratio columns
ratio_cols = [f'Ratio{i}' for i in range(1, 11)]

# Initialize StandardScaler
scaler = StandardScaler()

# Fit and transform the ratio columns
df[ratio_cols] = scaler.fit_transform(df[ratio_cols])

display(df.head())

,FYEAR,Bankruptcy Status,Ratio1,Ratio2,Ratio3,Ratio4,Ratio5,Ratio6,Ratio7,Ratio8,Ratio9,Ratio10
0,1999,0,0.476331,1.053185,1.735803,-0.275560,1.162526,-0.414404,-0.068379,1.263166,0.237498,0.002915
1,1999,0,0.856891,0.791236,0.645788,-0.397251,0.292861,-0.140721,0.443839,0.581442,0.987304,0.561897
2,1999,0,-0.212253,0.564539,0.505290,-0.154335,-0.300553,0.472220,-0.567606,0.623252,1.962386,2.051651
3,1994,1,-0.172429,0.594008,0.685079,-0.399773,-0.142138,0.586845,-0.473024,0.606309,0.118802,-0.094962
4,1999,0,1.916264,0.455371,0.399070,4.638573,-0.986969,-1.339275,2.934125,0.457907,-0.916468,0.559813


In [8]:
#the dataset is split into features (X) and target (y), where y is the bankruptcy status.
#Then, the data is divided into training and testing sets (80% train, 20% test) to train the model and evaluate its performance on unseen data.
#The shapes of the resulting sets are printed to verify the split.
X = df.drop('Bankruptcy Status', axis=1)
y = df['Bankruptcy Status']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Training set shape:", X_train.shape)
print("Testing set shape:", X_test.shape)

Training set shape: (2140, 11)
Testing set shape: (535, 11)


In [9]:
#the features (X) are explicitly selected as the 10 financial ratios (Ratio1 to Ratio10), and the target (y) is set as the Bankruptcy Status.
X = df[[f'Ratio{i}' for i in range(1, 11)]]
y = df['Bankruptcy Status']

display(X.head())
display(y.head())

,Ratio1,Ratio2,Ratio3,Ratio4,Ratio5,Ratio6,Ratio7,Ratio8,Ratio9,Ratio10
0,0.476331,1.053185,1.735803,-0.275560,1.162526,-0.414404,-0.068379,1.263166,0.237498,0.002915
1,0.856891,0.791236,0.645788,-0.397251,0.292861,-0.140721,0.443839,0.581442,0.987304,0.561897
2,-0.212253,0.564539,0.505290,-0.154335,-0.300553,0.472220,-0.567606,0.623252,1.962386,2.051651
3,-0.172429,0.594008,0.685079,-0.399773,-0.142138,0.586845,-0.473024,0.606309,0.118802,-0.094962
4,1.916264,0.455371,0.399070,4.638573,-0.986969,-1.339275,2.934125,0.457907,-0.916468,0.559813


,Bankruptcy Status
0,0
1,0
2,0
3,1
4,0


In [10]:
#a logistic regression model is initialized and then trained (fitted) on the training data (X_train and y_train)
# to learn the relationship between the financial ratios and bankruptcy status.
model = LogisticRegression()
model.fit(X_train, y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [12]:
#the coefficients of the trained logistic regression model are extracted and
#stored in a DataFrame alongside their corresponding features.
#Displaying this table helps to interpret the effect of each financial ratio on the probability of bankruptcy.

coefficients = pd.DataFrame({'Feature': X_train.columns, 'Coefficient': model.coef_[0]})
display(coefficients)
#(β values for each ratio + intercept).

#Positive β → increases bankruptcy odds.

#Negative β → decreases bankruptcy odds.

,Feature,Coefficient
0,FYEAR,-0.001417
1,Ratio1,0.152255
2,Ratio2,0.701456
3,Ratio3,-0.546670
4,Ratio4,-0.697227
5,Ratio5,0.163983
6,Ratio6,0.247299
7,Ratio7,-0.485685
8,Ratio8,-0.374727
9,Ratio9,0.148976


Most influential positive predictors (increase bankruptcy risk):

Ratio2 > Ratio6, Ratio5, Ratio1, Ratio9

Most protective features (reduce bankruptcy risk):

Ratio10 (strongest) > Ratio4, Ratio3, Ratio7, Ratio8

FYEAR has negligible effect — no need to include as a predictor unless analyzing trends over time.

In [13]:
y_pred = model.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
display(cm)

array([[441,  18],
       [ 48,  28]])

TP (True Positive) = 28 → correctly predicted bankrupt firms.

TN (True Negative) = 441 → correctly predicted non-bankrupt firms.

FP (False Positive) = 18 → predicted bankruptcy but actually not.

FN (False Negative) = 48 → missed bankrupt firms (predicted non-bankrupt).

In [14]:
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.8766355140186916

Classification Report:
               precision    recall  f1-score   support

           0       0.90      0.96      0.93       459
           1       0.61      0.37      0.46        76

    accuracy                           0.88       535
   macro avg       0.76      0.66      0.69       535
weighted avg       0.86      0.88      0.86       535



High accuracy (88%) is mostly driven by the non-bankrupt majority class.

Bankrupt class performance is weak:

Recall = 37% → model misses ~63% of actual bankruptcies → risky for financial risk prediction.

Precision = 61% → moderate → when model predicts bankruptcy, it is correct ~61% of the time.

F1 = 0.46 → overall weak performance for the minority class.

Non-bankrupt class performance is strong → model favors the majority class.

Class imbalance is the main issue → there are only 76 bankrupt cases vs 459 non-bankrupt.

In [15]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

#Calculate VIF (Variance Inflation Factor) for each ratio to detect multicollinearity.
# Since we already defined X with only the ratio columns, we can directly calculate VIF
#Displaying VIF helps identify which ratios may need removal or regularization.

vif_data = pd.DataFrame()
vif_data["feature"] = X.columns
vif_data["VIF"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

display(vif_data)

,feature,VIF
0,Ratio1,3.719746
1,Ratio2,3.029535
2,Ratio3,5.353898
3,Ratio4,2.117477
4,Ratio5,1.622126
5,Ratio6,2.702112
6,Ratio7,2.952201
7,Ratio8,4.842584
8,Ratio9,5.525317
9,Ratio10,4.235696


In [16]:
#A logistic regression model with class balancing (class_weight='balanced') is initialized to give more weight to the minority class (bankrupt firms).
model_balanced = LogisticRegression(max_iter=1000, class_weight='balanced')
model_balanced.fit(X_train, y_train)

y_pred_balanced = model_balanced.predict(X_test)

print("Accuracy (Balanced):", accuracy_score(y_test, y_pred_balanced))
print("\nClassification Report (Balanced):\n", classification_report(y_test, y_pred_balanced))

Accuracy (Balanced): 0.8523364485981308

Classification Report (Balanced):
               precision    recall  f1-score   support

           0       0.97      0.85      0.91       459
           1       0.49      0.84      0.62        76

    accuracy                           0.85       535
   macro avg       0.73      0.85      0.76       535
weighted avg       0.90      0.85      0.87       535



/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Recall for bankrupt class jumped from 37% → 84% ✅

Now the model correctly identifies most bankrupt firms, which is critical in risk prediction.

Precision dropped from 61% → 49%

Some more false positives (non-bankrupt predicted as bankrupt).

Acceptable trade-off: in bankruptcy prediction, catching bankrupts is more important than false alarms.

Non-bankrupt class performance slightly decreased:

Recall dropped from 96% → 85%

Still acceptable, but some non-bankrupts are now predicted as bankrupt due to balancing.

Using class_weight='balanced' successfully improved recall for the minority class.

This is more suitable for financial risk applications, where missing a bankrupt firm is costly.

Accuracy dropped slightly, but recall improvement for bankrupts is far more important.

In [17]:
# Apply L1 (Lasso) regularization
model_l1 = LogisticRegression(penalty='l1', solver='liblinear', max_iter=1000, class_weight='balanced')
model_l1.fit(X_train, y_train)

# Evaluate the model with L1 regularization
y_pred_l1 = model_l1.predict(X_test)

print("Accuracy (L1 Regularization):", accuracy_score(y_test, y_pred_l1))
print("\nClassification Report (L1 Regularization):\n", classification_report(y_test, y_pred_l1))

# Display coefficients with L1 regularization
coefficients_l1 = pd.DataFrame({'Feature': X_train.columns, 'Coefficient': model_l1.coef_[0]})
display(coefficients_l1)

Accuracy (L1 Regularization): 0.7813084112149533

Classification Report (L1 Regularization):
               precision    recall  f1-score   support

           0       0.96      0.78      0.86       459
           1       0.37      0.79      0.51        76

    accuracy                           0.78       535
   macro avg       0.66      0.78      0.68       535
weighted avg       0.87      0.78      0.81       535



/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


,Feature,Coefficient
0,FYEAR,-0.004500
1,Ratio1,0.205042
2,Ratio2,0.707808
3,Ratio3,-0.570378
4,Ratio4,-0.249259
5,Ratio5,0.117675
6,Ratio6,0.368944
7,Ratio7,-0.539964
8,Ratio8,-0.170886
9,Ratio9,0.035203


Overall:

Accuracy: 0.78 → lower than previous models (expected because L1 shrinks coefficients and removes minor features).

Recall for bankrupt class: 79% → still very high ✅ (good for catching bankrupt firms).

Precision for bankrupt class: 37% → dropped, more false positives (trade-off for high recall).

Weighted F1: 0.81 → moderate.

Observations:

L1 shrank low-impact coefficients (e.g., Ratio9 is near zero).

Most important predictors:

Positive: Ratio2, Ratio6, Ratio1

Negative (protective): Ratio10, Ratio3, Ratio7

Interpretability improved → you can clearly see which ratios matter most for bankruptcy.

Key Takeaways

Recall improved for bankrupt class → high-risk firms are mostly detected.

Some precision loss → more false positives, acceptable in financial risk context.

Feature selection via L1 → reduces noise and highlights critical ratios: Ratio2, Ratio10, Ratio3, Ratio7.

Key Insights

High-Risk Indicators:

Firms with high Ratio2 and Ratio6 values should be closely monitored.

These ratios are the strongest predictors of bankruptcy risk.

Protective Indicators:

Ratio10, Ratio3, Ratio7 reduce bankruptcy risk → best practices associated with these ratios can be encouraged across the portfolio.

Actionable Recommendations:

Integrate these key ratios into early warning dashboards.

Prioritize high recall detection: catching potential bankrupt firms early is more critical than occasional false positives.

Review policies, operational metrics, or financial controls that influence Ratio2 and Ratio6.

Benchmark best practices from firms with high Ratio10, Ratio3, Ratio7.